### PR TITLE
Implement partial derivatives with boundary values

### LINKED ISSUE #200: Add a FDM partial operator with Dirichlet boundary conditions
We may need to have a centred FDM operator of order two which can handle boundary conditions, starting with Dirichlet ones

### PR DESCRIPTION
**Pull request by protaisM**
Resolve #200

### CODE REVIEW

**Comment by EmilyBourne** [comment]
> I am confused by the fact that that this PR triggers cpp warnings for src/pde_solvers/polarpoissonlikesolver.hpp...

If you haven't touched the file then you probably just need to update your branch. The test compares your branch to main so if a file has changed in main since you created the PR then it will be marked as relevant

File: src/math_tools/central_fdm_partial_derivatives_with_boundary_values.hpp [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+    using typename base_type::DFieldType;
+
+    /// The type of a constant reference to the field to be differentiated.
+    using typename base_type::DConstFieldType;
+
+    /// The index range of the dimension on which the partial derivative is calculated.
+    using typename base_type::IdxRangeDeriv;
+
+    /// The index range of all dimensions except DerivativeDimension.
+    using typename base_type::IdxRangeBatch;
```
Comment (EmilyBourne) [inline src/math_tools/central_fdm_partial_derivatives_with_boundary_values.hpp:41]: FYI you only need to give detailed comments for public attributes.

That said I'm not sure why these are private. I'm not sure nvidia likes it when the types used in the function prototypes are private

File: src/math_tools/central_fdm_partial_derivatives_with_boundary_values.hpp [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+class CentralFDMPartialDerivativeWithBValueCreator
+    : public IPartialDerivativeCreator<IdxRangeFull, DerivativeDimension>
+{
+private:
+    /// The type of a constant reference to the field to be differentiated.
+    using DConstFieldType = DConstField<IdxRangeFull>;
+
+    // The boundary values for the derivative
+    double m_BValue_left;
+    double m_BValue_right;
```
Comment (EmilyBourne) [inline src/math_tools/central_fdm_partial_derivatives_with_boundary_values.hpp:177]: As above, snake_case not respected

File: src/math_tools/central_fdm_partial_derivatives_with_boundary_values.hpp [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+ */
+template <class IdxRangeFull, class DerivativeDimension>
+class CentralFDMPartialDerivativeWithBValueCreator
+    : public IPartialDerivativeCreator<IdxRangeFull, DerivativeDimension>
+{
+private:
+    /// The type of a constant reference to the field to be differentiated.
+    using DConstFieldType = DConstField<IdxRangeFull>;
+
+    // The boundary values for the derivative
```
Comment (EmilyBourne) [inline src/math_tools/central_fdm_partial_derivatives_with_boundary_values.hpp:175]:
```suggestion
public:
    /// The type of a constant reference to the field to be differentiated.
    using DConstFieldType = DConstField<IdxRangeFull>;

private:
    // The boundary values for the derivative
```

File: tests/math_tools/test_partial_derivatives.cpp [hunk 4]
```diff
@@ -98,6 +99,73 @@ class FunctionToDifferentiateCosine
     }
 };
 
+/**
+ * @brief A class that represents a test function for computing partial derivatives with boundary values.
+ * The test function is defined as the product of two polynomials of cosine functions.
+ * This function has been chosen to ease the use of boundary conditions
+ * due to the fact that its value and its derivative are both null at the boundary.
+ */
+class FunctionToDifferentiateLinearCosine
```
Comment (EmilyBourne) [inline tests/math_tools/test_partial_derivatives.cpp:108]: Why do you need a new test function? Should we test all the methods with both test functions?

File: tests/math_tools/test_partial_derivatives.cpp [hunk 5]
```diff
@@ -519,6 +587,67 @@ class PartialDerivativeTestFDM : public PartialDerivativeTest<ncells_x, ncells_y
...
+        IdxRangeY idxrange_y(typename base_type::IdxY(0), base_type::m_ncells_y + 1);
+        IdxRangeXY idxrange_xy(idxrange_x, idxrange_y);
+
+        // the boundary values are the value of the test function at -pi and pi, namely 0.
+        CentralFDMPartialDerivativeWithBValueCreator<IdxRangeXY, DerivativeDimension> const
+                derivative_creator(0., 0.);
+
+        FunctionToDifferentiateLinearCosine function_to_differentiate;
+        double const max_error = base_type::template compute_max_error<
+                DerivativeDimension,
```
Comment (EmilyBourne) [inline tests/math_tools/test_partial_derivatives.cpp:641]: Do you need to do this? Does the following not compile:
```suggestion
        double const max_error = compute_max_error<
                DerivativeDimension,
```
?
It seems simpler but it is possible that the templates mess this up

**Review by EmilyBourne — CHANGES_REQUESTED** [review]

File: src/math_tools/central_fdm_partial_derivatives_with_boundary_values.hpp (hunk 3 above)
Comment (protaisM) [reply src/math_tools/central_fdm_partial_derivatives_with_boundary_values.hpp:175]: I based my private/public choices on the already implemented versions of partial derivatives. I don't care if our types are public or private, however I think we need to be consistent. Or maybe there is a reason to ask me to change the types for public visibility whereas 
- central_fdm_partial_derivatives.hpp
- gradient.hpp
- spline_1d_partial_derivative.hpp
- spline_2d_partial_derivative.hpp
- spline_builder_2d_cache.hpp
have their types in private visibility.

Comment (EmilyBourne) [reply src/math_tools/central_fdm_partial_derivatives_with_boundary_values.hpp:175]: Huh, I didn't notice it there. In that case I agree this is fine.

File: src/math_tools/central_fdm_partial_derivatives_with_boundary_values.hpp [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+    /// The index range of the dimension on which the partial derivative is calculated.
+    using typename base_type::IdxRangeDeriv;
+
+    /// The index range of all dimensions except DerivativeDimension.
+    using typename base_type::IdxRangeBatch;
+
+    /// The field to be differentiated
+    DFieldMemType m_field;
+    double m_BValue_left;
+    double m_BValue_right;
```
[the remark this answers is not in the document]
Comment (protaisM) [reply src/math_tools/central_fdm_partial_derivatives_with_boundary_values.hpp]: Fixed 325000e62233d3be2da126a76383c3c543be79c4

File: src/math_tools/central_fdm_partial_derivatives_with_boundary_values.hpp (hunk 2 above)
Comment (protaisM) [reply src/math_tools/central_fdm_partial_derivatives_with_boundary_values.hpp]: Fixed 325000e62233d3be2da126a76383c3c543be79c4

File: tests/math_tools/test_partial_derivatives.cpp (hunk 4 above)
Comment (protaisM) [reply tests/math_tools/test_partial_derivatives.cpp]: Fixed 325000e62233d3be2da126a76383c3c543be79c4

**Comment by EmilyBourne** [comment]
Feel free to ignore the coverage. It seems to be a collection bug rather than actual missing coverage

**Comment by protaisM** [comment]
I am unable to see the two lines of this MR which are not covered by the tests

**Comment by EmilyBourne** [comment]
> I am unable to see the two lines of this MR which are not covered by the tests

You can get there by going to:
https://app.codecov.io/gh/gyselax/gyselalibxx

picking the branch and navigating to the right file but the links should start working properly once my bug fixes go through in #243

File: tests/math_tools/test_partial_derivatives.cpp [hunk 7]
```diff
@@ -68,7 +69,7 @@ class FunctionToDifferentiateCosine
     {
         double const x = ddc::get<X>(coord_xy);
         double const y = ddc::get<Y>(coord_xy);
-        return Kokkos::cos(x) * Kokkos::cos(y);
+        return (Kokkos::cos(x) + 1) * (Kokkos::cos(y) + 1);
```
Comment (EmilyBourne) [inline tests/math_tools/test_partial_derivatives.cpp:72]: Can you explain this change? Is it for boundary conditions?

**Review by EmilyBourne — APPROVED** [review]
Looks good to me

File: tests/math_tools/test_partial_derivatives.cpp (hunk 7 above)
Comment (protaisM) [reply tests/math_tools/test_partial_derivatives.cpp:72]: I changed the test function to use a function with zero value at the boundary (if we take a domain [-pi,pi]x[-pi,pi]). It doesn't change the other tests but on this domain we know the boundary value, which is constant.

Comment (protaisM) [reply tests/math_tools/test_partial_derivatives.cpp]: Fixed 742515bd63204c78549eeaf0e218b19be8a86b88v

**Review by EmilyBourne — APPROVED** [review]
